bed_mesh: Mesh Bed Leveling for Klipper

Arksine opened the pull request:
This pull request adds a configurable implementation of Mesh Bed Leveling for Klipper.    This is accomplished through probing a rectangular grid, interpolating a mesh from the probed points, and adding a z-adjustment to moves based on their X,Y position relative to the mesh.  By default z-adjustment is phased out, starting at a layer height of 1mm and finishing at 10mm.  These values are configurable and phasing out can be disabled if desired.

Generally bed_mesh.py can be divided into 5 sections:
1) Helper Functions:
Basic helpers that don't necessarily belong in any of the classes listed below.  Some of them could potentially be moved elsewhere.
2) BedMesh class:
Entry point class created when [bed_mesh] is found in the config.  Registers itself as a gcode transform object and is responsible for applying transforms.  
3) BedMeshCalibrate class:
Handles probing functionality.  Generates points using min_point, max_point, and probe_count.  This method makes the configuration much smaller if you are probing a large number of points.
4) MoveSplitter class:
Traverses the next move looking for a desired z delta in the mesh.  If the delta is found, the move is split and sent to the toolhead with a z-offset added.  This process continues until the original move is completely processed.  Moves smaller than the move_check_distance parameter don't get traversed, they just get the appropriate z-offset added to the end of the move.
5) ZMesh class:
Interpolates and stores the mesh.  Supports lagrange polynomial interpolation and cardinal spline interpolation.  Lagrange works well for smaller samples (3x3 grid), but can produce strange results as the number of samples increase.  ZMesh also handles z value lookups.


Signed-off-by: Eric Callahan <arksine.code@gmail.com>

Arksine commented:
Some additional notes:

I'm not sure how useful the current implementation will be for delta printers.  I assume that they would require a mesh that is not rectangular.

Things not yet implemented, but planned:
* Mapping generated points to the probe location instead of the nozzle location.  This seems to be an area of confusion for some people.  Also boundary checking to make sure that the probe doesn't move off of the configured print area.
* Probing in the reverse direction.
* A manual override to deal with spots not picked up by the mesh or outside of the mesh.  This would essentially be a secondary user generated mesh that overlays the original.  I'm told Marlin does something like this, but I haven't had an opportunity to check.

I chose not to implement any of the above because I thought it would be a better idea to submit as is, let you review it and suggest changes, then add additional functionality later.

KevinOConnor commented on klippy/extras/bed_mesh.py:
```diff
@@ -0,0 +1,617 @@
+                logging.debug(
+                    "bed_mesh: Using Z position_endstop for z_offset: %.4f"
+                    % (self.z_offset))
+        # Register MESH_BED_LEVING command
```
Stale comment?

Arksine replied:
Oops, I changed the name of the module and forgot to change that comment.

KevinOConnor commented on klippy/extras/bed_mesh.py:
```diff
@@ -0,0 +1,617 @@
+            self.z_adjust = 0.
+            self.toolhead.move(newpos, speed)
+        else:
+            prev_pos = self.get_position()
```
FYI, the previous position will always be the last position sent to move() or the last position returned by get_position().  The toolhead and gcode classes cache last position as an optimization - so it is safe to cache it in the move transform class as well.

Arksine replied:
Okay, I can do that.

KevinOConnor commented on klippy/extras/bed_mesh.py:
```diff
@@ -0,0 +1,617 @@
+                        self.z_offset = next_z
+                        return self.current_pos[0], self.current_pos[1], \
+                            self.current_pos[2] + self.z_offset, \
+                            self.current_pos[3]
```
Interesting.  I would have done this a different way - I would have implemented the mesh as a triangular mesh, and then always done linear interpolation between mesh points and always do line splitting at the point a move crosses a triangle border.  The "triangular mesh" (a series of equal sized triangles evenly spaced throughout the bed) could be filled directly by probing each triangle point or by using the bicubic/lagrange algos on a coarsely probed grid.

I suspect the method used here of splitting moves heuristically may result in some quirks.  For example, if a user does a printing move from 0,0 => 10,0 and then later does a non printing move from -5,5 => 5,5 then the toolhead may not be at the same Z position when it crosses the line at position 0,5.  (Using linear interpolation within a triangular mesh wouldn't have this quirk.)

That said, I don't know how other firmwares implement mesh bed leveling, and I don't know if the triangular method would have other quirks.  So, this is just a high level comment.

Arksine commented on klippy/extras/bed_mesh.py:
```diff
@@ -0,0 +1,617 @@
+# Mesh Bed Leveling
+#
+# Copyright (C) 2018  Kevin O'Connor <kevin@koconnor.net>
```
I initially based this off of your bed_tilt code, so I thought it would be a good idea to keep you in the copyright notice.

Arksine commented on klippy/extras/bed_mesh.py:
```diff
@@ -0,0 +1,617 @@
+        # Register G81 as alias
+        self.gcode.register_command(
+            'G81', self.cmd_BED_MESH_OUTPUT,
+            desc=self.cmd_BED_MESH_OUTPUT_help)
```
Actually you raise a good point and I was questioning if I should remove G80 and G81 altogether, or add aliases for other firmwares.  I think your approach of using gcode_macro is best, so I will remove these.

Arksine commented on klippy/extras/bed_mesh.py:
```diff
@@ -0,0 +1,617 @@
+            x, y, z, e = self.toolhead.get_position()
+            return [x, y, z - self.z_adjust, e]
+    def move(self, newpos, speed):
+        gc_z = newpos[2] - self.gcode.base_position[2]
```
I subtract the base position so I can access the actual layer height of the gcode.  Perhaps there is a better way for me to track this.  If, for example, someone is using z_adjust by a significant number (say 1mm) fade wouldn't work correctly.  However there may be a better way for me to track layer height.  I'll give it some thought and come up with something.

KevinOConnor replied:
Ah - `newpos[2]` should be the actual layer height.  The gcode.py code is tasked with undoing any arbitrary transformations prior to calling `move_with_transform()`.

Arksine commented on klippy/extras/bed_mesh.py:
```diff
@@ -0,0 +1,617 @@
+        self.probe_params['max_y'] = max(points, key=lambda p: p[1])[1]
+        offset = parse_pair(config, ('probe_offset',))
+        self.probe_params['x_offset'] = offset[0]
+        self.probe_params['y_offset'] = offset[1]
```
That sounds good.  It shouldn't be any problem to move them after merging.

Arksine commented on klippy/extras/bed_mesh.py:
```diff
@@ -0,0 +1,617 @@
+            self.z_offset = self.get_probed_position()[2]
+        self.probe_helper.start_probe()
+        if self.end_gcode is not None:
+            self.gcode.run_script_from_command(self.end_gcode)
```
Sure, sounds good.  I'll remove this.

Arksine commented on klippy/extras/bed_mesh.py:
```diff
@@ -0,0 +1,617 @@
+        else:
+            # return current position minus the current z-adjustment
+            x, y, z, e = self.toolhead.get_position()
+            return [x, y, z - self.z_adjust, e]
```
Ah, I didn't think of that.  I can look it up in the mesh instead.

Arksine commented on klippy/extras/bed_mesh.py:
```diff
@@ -0,0 +1,617 @@
+                    self.toolhead.move(split_move, speed)
+                else:
+                    self.gcode.respond_error(
+                        "Mesh Leveling: Error splitting move ")
```
Okay, no problem.

Arksine commented:
I have pushed an updated version with the suggested fixes added, and my initial testing indicates that everything works well.
Changes:
* Removed the base_position references 
* Moved the code that determines the z-factor out of the MoveSplitter class and into BedMesh so that it can easily be used to calculate the z_offset for both get_position() and move()
* Raised self.gcode.error() in place of calling self.gcode.respond_error()
* Removed G80 and G81 aliases
* Removed end_gcode: configuration parameter
* Updated Docs to reflect above changes.

I noticed that there is already a pull request to update probe.py to include offsets.  The only addition I would need is a function to fetch the offsets so I can use them to build the mesh.   It wouldn't be any problem for me to add it after the fact, or update probe.py for the whole thing,  just don't want to step on anyone's toes.

Eric

KevinOConnor commented on klippy/extras/bed_mesh.py:
```diff
@@ -0,0 +1,596 @@
+        # Automatic probe:z_virtual_endstop XY detection
+        detect_offset = config.getboolean('detect_z_offset', False)
+        self.z_offset = None
+        if not detect_offset and config.has_section('stepper_z'):
```
One additional comment - I don't think mesh bed leveling needs to do z_offset detection - the z_offset passed to finalize() should be the only z_offset needed (the value passed to finalize() is the nominal distance between the nozzle and bed when the probe triggers).

I suspect this code is derived from the bed_tilt.py code - but that code is trying to do something a bit weird.  If bed_tilt is used in combination with a "probe:z_virtual_endstop" then there is a feedback between the calculated tilt and the XY position where Z homing is performed - the bed_tilt.py code does "back flips" to handle that so the user can make an appropriate update to their printer.cfg.  I don't think this applies to bed mesh (the user is not expected to update their config with the results of the mesh).

-Kevin

Arksine commented:
I did derive offset_detection from bed_tilt, my reasoning for keeping it was to eliminate a situation where the user inadvertently entered the incorrect z_offset under [probe], as in most cases its value should be the same as the Z position_endstop.  

I have removed detection, however I keep a check against the [steppper_z] position_endstop if it exists.  If its different from the z_offset passed to finalize() I log it and print a warning to the Terminal.

KevinOConnor commented:
Thanks!

FYI, the warning would only make sense if "probe:z_virtual_endstop" is in use.  (When using a traditional z endstop, it would be unusual for the stepper_z position_endstop to be the same as the probe's z_offset.)

-Kevin

oderwat commented:
Very nice work. But wouldn't it make much more sense that the bed mesh can / is be stored? Creating a mesh before every single print seems to be a great waste of time to me. The bed should not change its shape so often.

Arksine commented:
@KevinOConnor Okay, I'll update the check to make sure that the probe is being used as the endstop.

@oderwat It depends on the bed and material.  The Prusa i3 MK3 has an aluminum bed with a removable steel sheet, so the change in geometry is not entirely predictable with heat, thus you really need to probe before every print.  I can see how other beds may not suffer from this issue and it would be desirable to be able to store and load a mesh.  Several people have requested this so I'll add it, it seems fairly trivial to accomplish, as it can be done with two gcodes:

`BED_MESH_SAVE FILE=<file path>
BED_MESH_LOAD FILE=<file path>`

Eric

KevinOConnor commented:
Ultimately, I think we'll need to come up with a general mechanism for saving and restoring calibration parameters.  The PID_CALIBRATE, DELTA_CALIBRATE, BED_TILT_CALIBRATE, and now BED_MESH_CALIBRATE all really want to an automatic mechanism for saving and restoring state.

I recently ran into this problem in my attempt to enhance DELTA_CALIBRATE - in order to perform an extended 9-point calibration I really need to save/restore the raw measurements, but asking the user to do that manually is way too onerous.

This was briefly discussed recently in issue #507.  I'm not really sure what the best way to proceed is.

-Kevin

w1ebr commented:
@KevinOConnor, what if printer.config supported an "include_calibration_data filename" directive? The software generated  version of the calibration data could be "printer.caldata" or something like that. The format of the file should be like the existing printer.config file. This would support one having more than one calibration file if they want to create one or edit a a software generated version.

Arksine commented:
My initial thought is to create a new module called storage.py.  It wouldn't need config (unless you want to make the file name/location configurable) so it could be registered after gcode in klippy.py.

When storage.py initializes it reads a file from disk (if it exists, otherwise just creates a file) and loads a dictionary containing all persistent data.   If a module wants access to persistent data, it would register itself with storage.py and provide two callbacks, on_load() and on_store().

Storage.py would register two gcodes:
`STORE MODULE=bed_mesh KEY=60`
Executes the on_store() callback of the module, then stores the returned data to the 'bed_mesh_60' key  in the dictionary.  After a storage call the dictionary is serialized (json or pickle) and written to disk.

`LOAD MODULE=bed_mesh KEY=60`
Retrieves data from the 'bed_mesh_60' key in the persistent dictionary and loads it via bed_mesh's on_load() callback.

This is sort of a brute force way to do this, but to me its the most obvious way to get the job done without breaking stuff.  Kevin, if you would like I could proceed to put this together.

bruce356 commented:
@Arksine , hi, please try to accommodate the saving of multiple mesh data (borosilicate glass bed) for varying temperatures of the bed. 
Cold bed for flexibles
60 deg for PLA
90 deg for Petg
110 deg for ABS
etc.
As the temperature of the bed changes the profile of the glass surface also changes but it remains the same for each temperature used so probing the bed each time before a print is time wasting. 
Thanks & regards - bruce

bruce356 commented:
Just wondering why G80 and G81 have been removed its a lot easier to remember than for example BED_MESH_CALIBRATE, for me anyway, could we not have both? its obviously Kevin's decision.
thanks - bruce

oderwat commented:
@bruce356 You easily can create a GCode Macro for that. But afaik you can't use parameters which such macros. So it is limited to commands which don't have them.
